File: Algorithms/Sorting-Algorithms/Quick-Sort/quick-sort.py
# Quick Sort: Select a pivot randomly, place the pivot in it's correct position while simultaneously,
# sorting/storing smaller (than or equal to) elements on the left and larger elements on the right. 

# TC: O(n*log(n))   -> log to the base 2, n
# SC: O(1)          -> Not counting recursion stack space 



def swap(arr: list[int], i: int, j: int) -> None:
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
    
    
def partition(arr: list[int], low: int, high: int) -> int:
    pivot = arr[low]
    i = low
    j = high
    while i<j:
        while arr[i]<=pivot and i<high:
            i+=1
        while arr[j]>pivot and j>low:
            j-=1
        if i<j:
            swap(arr, i, j)
    swap(arr, low, j)
    return j


def quick_sort(arr: list[int], low: int, high: int) -> None:
    if low < high:
        pivot = partition(arr, low, high)
        quick_sort(arr, low, pivot-1)
        quick_sort(arr, pivot+1, high)


def create_arr() -> list[int]:
    n = int(input("Enter the number of elements in the array: "))
    arr = []
    print("Enter the elements of array: ")
    for i in range(n):
        arr.append(int(input()))
    print()
    print("The array is: ")
    for i in arr:
        print(i, end=" ")
    print()
    return arr


if __name__ == "__main__":
    arr = create_arr()
    n = len(arr)
    quick_sort(arr, 0, n-1)
    print("Sorted Array: ")
    for i in arr:
        print(i, end=" ")
    print()

File: Algorithms/Sorting-Algorithms/Quick-Sort/quick-sort.cpp
#include<iostream>
#include<vector>
using namespace std;

void swap1(int &a, int &b) {
    int temp = a;
    a = b;
    b = temp;
}


vector<int> form_vec() {
    int n;
    cout<<"Enter the number of elements of the array: ";
    cin >> n;
    vector<int> arr(n);
    cout<<"Enter the elements of the array: "<<endl;
    for(int i=0; i<n; i++)
        cin >> arr[i];
    cout<<endl;
    cout<<"The array is: "<<endl;
    for(int i: arr)
        cout<< i <<" ";
    cout<<endl<<endl;
    return arr;
}


int partition(vector<int> &arr, int low, int high) {
    int pivot = arr[low];
    int i = low;
    int j = high;
    while(i<j) {
        while(arr[i]<=pivot && i<high)
            i++;
        while(arr[j]>pivot && j>low)
            j--;
        if(i<j)
            swap1(arr[i], arr[j]);
    }
    swap1(arr[low], arr[j]);
    return j;
}


void quick_sort(vector<int> &arr, int low, int high) {
    if(low < high) {
        int pivot = partition(arr, low, high);
        quick_sort(arr, low, pivot-1);
        quick_sort(arr, pivot+1, high);
    }
}


void quick_sort_decode(vector<int> &arr, int low, int high) {
    if(low < high) {
        cout<<"Array before pivoting: \n";
        for(int i: arr)
            cout<<i<<"    ";

        int pivot = partition(arr, low, high);

        cout<<"\nArray after placing pivot(1st) element in correct position: \n";
        for(int i: arr)
            cout<<i<<"    ";
        cout<<"Correct / Pivot position: "<<pivot<<"\n\n";

        quick_sort_decode(arr, low, pivot-1);
        quick_sort_decode(arr, pivot+1, high);
    }
}



int main() {
    vector<int> arr = form_vec();
    int n = arr.size();
    quick_sort(arr, 0, n-1);
    cout<<"Sorted Array: "<<endl;
    for(int i: arr)
        cout<< i <<" ";
    cout<<endl;
    return 0;

}